refactor: refactor shutdown logic in curp

themanforfree opened the pull request:
Please briefly answer these questions:
#395 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

iGxnon commented:
tokio has a channel that can meet the requirements: `tokio::sync::watch::channel`
```rust
use std::time::Duration;
use tokio::select;
use tokio::sync::watch;
use tokio::time::sleep;

#[tokio::main]
async fn main() {
    let (shutdown_trigger, mut rx_1) = watch::channel(());

    let mut rx_2 = rx_1.clone();

    tokio::spawn(async move {
        let _ = rx_1.changed().await;
        println!("Do some finishing work.");
        drop(rx_1); // it is ok to omit this line
    });

    tokio::spawn(async move {
        let _ = rx_2.changed().await;
        println!("Do some finishing work.");
    });

    sleep(Duration::from_secs(3)).await;

    println!("I want to shut down my tasks");
    let _ = shutdown_trigger.send(());

    shutdown_trigger.closed().await;
    println!("I have done all finishing works");
}
```
It looks like you have already reinvent the wheel in `utils::shutdown` :D

themanforfree commented:
> tokio has a channel that can meet the requirements: `tokio::sync::watch::channel`
> 
> ```rust
> use std::time::Duration;
> use tokio::select;
> use tokio::sync::watch;
> use tokio::time::sleep;
> 
> #[tokio::main]
> async fn main() {
>     let (shutdown_trigger, mut rx_1) = watch::channel(());
> 
>     let mut rx_2 = rx_1.clone();
> 
>     tokio::spawn(async move {
>         let _ = rx_1.changed().await;
>         println!("Do some finishing work.");
>         drop(rx_1); // it is ok to omit this line
>     });
> 
>     tokio::spawn(async move {
>         let _ = rx_2.changed().await;
>         println!("Do some finishing work.");
>     });
> 
>     sleep(Duration::from_secs(3)).await;
> 
>     println!("I want to shut down my tasks");
>     let _ = shutdown_trigger.send(());
> 
>     shutdown_trigger.closed().await;
>     println!("I have done all finishing works");
> }
> ```
> 
> It looks like you have already reinvent the wheel in `utils::shutdown` :D

My mistake, I know this channel, but I don’t know that it can wait for all `Receiver` to be drop

bsbds commented on curp/src/server/cmd_worker/conflict_checked_mpmc.rs:
```diff
@@ -546,6 +564,16 @@ pub(in crate::server) fn channel<C: 'static + Command, CE: 'static + CommandExec
             }
         }
+
+        // cleanup filter after receiving shutdown signal
```
why do we have to cleanup the filter instead of drop it directly?

bsbds commented on curp/src/server/cmd_worker/conflict_checked_mpmc.rs:
```diff
@@ -505,6 +505,20 @@ impl<C: Command, CE: CommandExecutor<C>> Filter<C, CE> {
+        self.vs.iter().all(|(_, v)| {
+            matches!(
+                v.inner,
+                VertexInner::Cmd {
```
could the vs contains some vertex whose type is not a Cmd, so that it will block the shutdown?

themanforfree replied:
Other vs can be executed normally to the end without waiting for a new CEEvent

themanforfree commented on simulation/src/curp_group.rs:
```diff
@@ -84,7 +85,7 @@ impl CurpGroup {
                 let role_change_cb = TestRoleChange::default();
                 let role_change_arc = role_change_cb.get_inner_arc();
-
+                let (t, l) = watch::channel(false);
```
Fixed

themanforfree commented on xline/src/server/xline_server.rs:
```diff
@@ -83,6 +89,7 @@ impl XlineServer {
         storage_config: StorageConfig,
         compact_config: CompactConfig,
     ) -> Self {
+        let (tx, _) = watch::channel(false);
```
Fixed

Phoenix500526 commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+    /// Check if the mpsc channel and sync follower daemon has been shutdown.
+    #[inline]
+    #[must_use]
+    pub fn check_partial_shutdown(&self) -> bool {
```
Why should we consider the scenario that both the `mpsc_change_shutdown` and the `sync_follower_daemon_shutdown` are true as a partial shutdown case?

themanforfree replied:
Follower only needs to ensure that there is no command being processed in mpmc channel, and Leader also needs to ensure that all logs has been synchronized to follower.

Phoenix500526 commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+
+    /// Send the shutdown signal and wait for all listeners drop.
+    #[inline]
+    pub async fn shutdown_and_wait(&self) {
```
I think that `self_shutdown` is a better name, because it's more straightforward.

themanforfree replied:
This structure also has a method called `shutdown`. I think that method should be renamed `self_shutdown`, and this method should be renamed `self_shutdown_and_wait`. They send the same signal, but behave differently.

Phoenix500526 commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+use tracing::warn;
+
+/// Shutdown Signal
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```
I think we don't need to derive `PartialEq` and `Eq` here.

themanforfree replied:
Fixed

Phoenix500526 commented on xline/src/server/xline_server.rs:
```diff
@@ -66,6 +70,8 @@ pub struct XlineServer {
     server_timeout: ServerTimeout,
     /// Shutdown trigger
     shutdown_trigger: Arc<Event>,
+    /// Curp shutdown trigger
```
Is it necessary to keep `shutdown_trigger` and `curp_shutdown_trigger` in `XlineServer`? Can we remove the `shutdown_trigger`?

themanforfree replied:
it is a temporary solution, this PR contains only the shutdown logic of curp

Phoenix500526 replied:
So we will finally remove the `shutdown_trigger` but not in this PR? If so, please add an issue to track it.

themanforfree replied:
#436

Phoenix500526 commented on curp/src/log_entry.rs:
```diff
@@ -24,6 +24,8 @@ pub(crate) enum EntryData<C> {
     Command(Arc<C>),
     /// `ConfChange` entry
     ConfChange(ProposeId),
+    /// `ConfChange` entry
```
Please modify the comment.

themanforfree replied:
Fixed

Phoenix500526 commented on curp/src/log_entry.rs:
```diff
@@ -39,6 +41,15 @@ where
         }
     }
 
+    /// Create a new `LogEntry` of a `ConfChange`
```
Same as above

themanforfree replied:
Fixed

Phoenix500526 commented on curp/src/client.rs:
```diff
@@ -353,6 +366,38 @@ where
         }
     }
 
+    /// The slow round of Curp protocol
```
Please modify the comment.

themanforfree replied:
FIxed

Phoenix500526 commented on curp/src/server/cmd_worker/conflict_checked_mpmc.rs:
```diff
@@ -59,7 +60,7 @@ pub(super) enum TaskType<C: Command> {
     SpecExe(Arc<LogEntry<C>>, Option<C::Error>),
     /// After sync a cmd
-    AS(Arc<LogEntry<C>>, C::PR),
+    AS(Arc<LogEntry<C>>, Option<C::PR>),
     /// Reset the CE
```
Why should we need to covert the `C::PR` into `Option<C::PR>`?

themanforfree replied:
Shutdown entry does not need C::PR

Phoenix500526 commented on curp/src/server/raw_curp/mod.rs:
```diff
@@ -280,6 +283,48 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
+        &self,
+        propose_id: ProposeId,
+    ) -> ((Option<ServerId>, u64), Result<(), ProposeError>) {
+        let st_r = self.st.read();
```
Please use `map_read` and fetch the `leader_id`, `term` and `role` out. We can reduce the size of the critical section.

themanforfree replied:
We need to lock `st` and `log` at the same time, Otherwise, the term used by `log.push_shutdown` may be dirty data.

themanforfree commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+    pub async fn wait_self_shutdown(&mut self) {
+        loop {
+            let _ig = self.listener.changed().await;
+            if Signal::SelfShutdown == *self.listener.borrow() {
```
Fixed

themanforfree commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+    /// Sender for shutdown signal.
+    trigger: watch::Sender<Signal>,
+    /// State of mpsc channel.
+    mpsc_change_shutdown: AtomicBool,
```
Fixed

themanforfree commented on curp/src/server/cmd_worker/mod.rs:
```diff
@@ -70,6 +70,9 @@ async fn cmd_worker<
     done_tx: flume::Sender<(Task<C>, bool)>,
     curp: Arc<RawCurp<C, RC>>,
     ce: Arc<CE>,
+    // This task will safely exit when the log_tx is dropped, but we still
```
Fixed, it should be dispatch_rx

iGxnon commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+        self.inner
+            .mpsc_change_shutdown
+            .store(true, Ordering::Relaxed);
+        if self.check_partial_shutdown() {
+            self.shutdown();
+        }
```
I think it's advisable to omit this piece of code. The name of this function doesn't explicitly indicate that it will send a shutdown signal.

themanforfree replied:
Fixed

iGxnon commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+        self.inner
+            .sync_follower_daemon_shutdown
+            .store(true, Ordering::Relaxed);
+        if self.check_partial_shutdown() {
+            self.shutdown();
+        }
```
Same as above

themanforfree replied:
Fixed

iGxnon commented on utils/src/shutdown.rs:
```diff
@@ -0,0 +1,163 @@
+/// Create a channel for shutdown.
+#[must_use]
+#[inline]
+pub fn channel() -> (Trigger, Listener) {
```
We can acquire Listeners through `Trigger::subscribe`, so could we only return the Trigger here?

themanforfree replied:
The internal watch channel is created in pairs, and I think it is better to give the upper layer the choice than to discard it directly.

iGxnon commented on curp/src/server/cmd_worker/conflict_checked_mpmc.rs:
```diff
@@ -541,25 +540,59 @@ pub(in crate::server) fn channel<C: 'static + Command, CE: 'static + CommandExec
+    loop {
+        tokio::select! {
+            biased; // cleanup filter first so that the buffer in filter can be kept as small as possible
+            sig = shutdown_listener.wait() => {
```
can we add `; if !is_shutdown_state` here to prevent a poll of shutdown_listener?

themanforfree replied:
Fixed

themanforfree commented on curp/src/server/curp_node.rs:
```diff
@@ -300,30 +331,56 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
             if !curp.is_leader() {
-                leader_event.listen().await;
+                shutdown_trigger.mark_sync_daemon_shutdown();
+                shutdown_trigger.check_and_shutdown();
```
Follower only needs to ensure that there is no command being processed in mpmc channel, and Leader also needs to ensure that all logs has been synchronized to follower.

bsbds replied:
But it seems after the follower becomes leader, the `sync_follower_daemon_shutdown` be `true` still holds.

themanforfree replied:
Fixed

themanforfree commented on curp/src/client.rs:
```diff
@@ -353,6 +366,38 @@ where
+    /// The shutdown rpc of curp protocol
+    #[instrument(skip_all)]
+    pub async fn shutdown(&self, id: ProposeId) -> Result<(), ProposeError> {
+        loop {
```
This already has it.

themanforfree commented on utils/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ categories = ["utilities"]
 full = ["std", "tokio", "parking_lot"]
 std = []
-tokio = ["dep:tokio", "dep:async-trait"]
+tokio = ["dep:async-trait"]
```
This feature is the utility of Tokio's lock, not the runtime, I don't think it's weird.

Phoenix500526 commented on curp/tests/common/curp_group.rs:
```diff
@@ -38,24 +49,45 @@ impl SnapshotAllocator for MemorySnapshotAllocator {
 }
 
+/// Rocks snapshot allocator
+pub(crate) struct RocksSnapshotAllocator;
```
Why should we re-define a `RocksSanpshotAllocator` here rather than reusing the `RocksSnapshotAllocator` defined in `snapshot_allocator.rs`?

themanforfree replied:
Here is `curp`'s Integration test, `RocksSnapshotAllocator` is defined in `xline` crate, `curp` should not depend on `xline`

themanforfree replied:
snapshot_allocator.rs is defined in xline, curps's tests should not depends on xline

Phoenix500526 replied:
But these two pieces of code are totally the same. Can we extract them into a shared part?

themanforfree replied:
Fixed

themanforfree commented on xline/src/server/lease_server.rs:
```diff
@@ -324,7 +328,11 @@ where
+                .client
+                .get_leader_id_from_curp()
+                .await
+                .map_err(|e| tonic::Status::internal(format!("Get leader id error: {e}")))?;
```
`propose_err_to_status` convert `CommandProposeError` to `tonic::Status`, not `ProposeError`

Phoenix500526 replied:
So why don't we convert the `ProposeError` to the `CommandProposeError` and then use `propose_err_to_status`?  IMO, our current error-handling logic is somewhat complex. It's better to unify it using a single method for error transformation, as it will be less prone to oversight during the subsequent refactoring process.

themanforfree commented on xlinectl/src/main.rs:
```diff
@@ -214,6 +214,11 @@ fn cli() -> Command {
             .help_heading(GLOBAL_HEADING)
             .value_parser(value_parser!(u64))
             .default_value("50"))
+        .arg(arg!(--retry_count <TIMEOUT> "The count of Curp client retry times")
```
Renamed

themanforfree commented on curp/src/client.rs:
```diff
@@ -295,10 +305,16 @@ where
         debug!("slow round for cmd({}) started", cmd.id());
         let retry_timeout = *self.timeout.retry_timeout();
-        loop {
+        let retry_count = *self.timeout.retry_count();
```
Renamed

themanforfree commented on curp/tests/server.rs:
```diff
@@ -254,5 +258,49 @@ async fn concurrent_cmd_order_should_have_correct_revision() {
+    let res = client
+        .propose(TestCommand::new_put(vec![888], 1), false)
+        .await;
+    println!("{:?}", res);
```
Just for test, removed

themanforfree commented on utils/src/config.rs:
```diff
@@ -406,6 +413,11 @@ pub struct ClientTimeout {
+    /// Curp client retry interval
+    #[getset(get = "pub")]
+    #[serde(default = "default_retry_count")]
+    retry_count: usize,
```
Renamed

themanforfree commented on curp/src/error.rs:
```diff
@@ -111,6 +120,9 @@ impl From<bincode::Error> for ProposeError {
 pub enum CommandProposeError<C: Command> {
+    /// Cluster already shutdown
+    #[error("cluster shutdown")]
+    Shutdown,
```
Fixed

bsbds commented on curp/tests/common/curp_group.rs:
```diff
@@ -38,24 +49,45 @@ impl SnapshotAllocator for MemorySnapshotAllocator {
+#[async_trait]
+impl SnapshotAllocator for RocksSnapshotAllocator {
+    async fn allocate_new_snapshot(&self) -> Result<Snapshot, Box<dyn Error>> {
+        let tmp_path = format!("/tmp/snapshot-{}", uuid::Uuid::new_v4());
```
Please consider use [tempfile](https://docs.rs/tempfile/latest/tempfile/#) for temp file creation.

themanforfree replied:
Fixed

bsbds commented on curp/tests/server.rs:
```diff
@@ -254,5 +261,52 @@ async fn concurrent_cmd_order_should_have_correct_revision() {
+    let res = client
+        .propose(TestCommand::new_put(vec![888], 1), false)
+        .await;
+    println!("*****{res:?}");
```
Please remove this debug print.

themanforfree replied:
Removed

bsbds commented on curp/src/server/curp_node.rs:
```diff
@@ -312,30 +340,58 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
+                        return;
+                    }
+                    _ = leader_event.listen() => {
+                        shutdown_trigger.mark_sync_daemon_shutdown();
```
This should be `reset_sync_daemon_shutdown`

themanforfree replied:
Fixed
